crates/bevy_ecs/src/world/interior_mutable_world.rs

```diff
@@ -0,0 +1,473 @@
+///     let interior_mutable_world = world.as_interior_mutable();
+///     let resource_access = OnlyResourceAccessWorld(interior_mutable_world);
+///     let component_access = OnlyComponentAccessWorld(interior_mutable_world);
+///     (resource_access, component_access)
+/// }
+/// ```
+#[derive(Copy, Clone)]
+pub struct InteriorMutableWorld<'w>(&'w World);
+
+impl<'w> InteriorMutableWorld<'w> {
+    pub(crate) fn new(world: &'w World) -> Self {
+        InteriorMutableWorld(world)
+    }
+
+    /// Gets a reference to the [`&World`](crate::world::World) this [`InteriorMutableWorld`] belongs to.
+    /// This can be used to call methods like [`World::read_change_tick`] which aren't exposed here but don't perform any accesses.
+    ///
+    /// **Note**: You *must not* hand out a `&World` reference to arbitrary safe code when the [`InteriorMutableWorld`] is currently
+    /// being used for mutable accesses.
+    ///
+    /// # Safety
+    /// - the world must not be used to access any resources or components. You can use it to safely access metadata.
+    pub unsafe fn world(&self) -> &'w World {
+        self.0
+    }
+
+    /// Retrieves this world's [Entities] collection
+    #[inline]
+    pub fn entities(&self) -> &'w Entities {
+        &self.0.entities
+    }
+
+    /// Retrieves this world's [Archetypes] collection
+    #[inline]
+    pub fn archetypes(&self) -> &'w Archetypes {
+        &self.0.archetypes
+    }
+
+    /// Retrieves this world's [Components] collection
+    #[inline]
+    pub fn components(&self) -> &'w Components {
+        &self.0.components
+    }
+
+    /// Retrieves this world's [Storages] collection
+    #[inline]
+    pub fn storages(&self) -> &'w Storages {
+        &self.0.storages
+    }
+
+    /// Retrieves this world's [Bundles] collection
+    #[inline]
+    pub fn bundles(&self) -> &'w Bundles {
+        &self.0.bundles
+    }
+
+    /// Retrieves an [`InteriorMutableEntityRef`] that exposes read and write operations for the given `entity`.
+    /// Similar to the [`InteriorMutableWorld`], you are in charge of making sure that no aliasing rules are violated.
+    pub fn get_entity(&self, entity: Entity) -> Option<InteriorMutableEntityRef<'w>> {
+        let location = self.0.entities.get(entity)?;
+        Some(InteriorMutableEntityRef {
+            world: InteriorMutableWorld(self.0),
+            entity,
+            location,
+        })
+    }
+
+    /// Gets a reference to the resource of the given type if it exists
+    ///
+    /// # Safety
+    /// All [`InteriorMutableWorld`] methods take `&self` and thus do not check that there is only one unique reference or multiple shared ones.
+    /// It is the callers responsibility to make sure that there will never be a mutable reference to a value that has other references pointing to it,
+    /// and that no arbitrary safe code can access a `&World` while some value is mutably borrowed.
```

Comment:
The safety comments on these methods have way too much irrelevant information, IMO. Correct me if I'm wrong, but the caller just needs to ensure that no mutable references to `R` can exist. Similarly with `InteriorMutableEntityRef` methods.